lib/conjur/authenticator.rb

```diff
@@ -0,0 +1,76 @@
+module Conjur
+  # Keeps a fresh Conjur access token in a named file by re-authenticating as needed.
+  class Authenticator
+    TOKEN_LIFESPAN = ( ENV['CONJUR_TOKEN_LIFESPAN'] || 5 * 60 ).to_i.seconds
+    DELAY = ( ENV['CONJUR_TOKEN_REFRESH_DELAY'] || 10 ).to_i.seconds
+
+    attr_reader :authenticate, :filename
+
+    # +authenticate+ should be a proc that authenticates with Conjur and returns an 
+    # access token as a Hash.
+    def initialize authenticate, filename
+      @authenticate = authenticate
+      @filename = filename
+    end
+
+    class << self
+      def default_filename
+        "/etc/conjur-auth/token"
+      end
+
+      # Check the token every +DELAY+ seconds and refresh it if it's out of date. 
+      def run authenticate:, filename: default_filename
+        while true
+          authenticator = Authenticator.new(authenticate, filename)
+          authenticator.refresh unless authenticator.fresh?
+          sleep DELAY
+        end
+      end
+    end
+
+    def fresh?
+      token && (token_age <= TOKEN_LIFESPAN)
+    end
+
+    # Perform atomic replacement of the token
+    def refresh
+      token = authenticate.call
+      require 'fileutils'
+      temp_filename = File.join(directory, "token.#{random}")
+      File.write temp_filename, JSON.pretty_generate(token)
+      FileUtils.mv temp_filename, filename
+      Conjur.log << "Refreshed Conjur auth token to #{filename.inspect}\n" if Conjur.log
+    rescue
+      $stderr.puts $!
```

Comment:
Should this use `warn` for clarity? Also, I note that errors don't go into log; I'd expect them to.